Rethink Acceptance Testing

rwjblue opened the pull request:
Continues the testing related RFC's (slowly chipping away at https://github.com/emberjs/rfcs/pull/119) by proposing migrating the acceptance testing API "unify" (though I'm not allowed to use that word in the title any more) with the newly landed and implemented ember-qunit API's (proposed in #232). 

[rendered](https://github.com/emberjs/rfcs/blob/acceptance-testing/text/0268-acceptance-testing-refactor.md)

simonihmig commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+helpers are being introduced to help avoid extra noise (e.g. "rightward shift")
+in tests.
+
+This means that users will be able to use either of the following interchangably:
```
@Do we really want/need two equivalent ways to use those helpers, or should'n there be one "right" way to do this? In the sense of strong Ember conventions, so teams don't end up bikeshedding about the right way. And maybe also useful for easier tooling (linting rules, codemods)? Just a thought...

rwjblue replied:
@simonihmig - Yes, the blueprints will use the imported helpers. The only reason I describe both is that the importable helpers _defer_ to the local versions on the test context (essentially as a desugaring).  Do you think I need to clarify that a bit?

rwjblue replied:
I updated some verbiage in this section to hopefully clarify things a bit better.

simonihmig commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+Here is a brief list of the more important but possibly understated changes
+being proposed here:
+
+* The global test helpers that exist now, will no longer be present (e.g.
```
Really not present, or rather deprecated? So they can co-exist, see the following section.

rwjblue replied:
When running tests with the old `moduleForAcceptance` API they will be present, this section is referring to tests using `setupAcceptanceTest(...)` API and is correct in saying that the global methods _will not_ be present.

simonihmig replied:
That makes sense! Maybe this should be stated to prevent misunderstandings like I had?

rwjblue commented:
> But could you maybe explain the relationship between this and #119? Does it supercede the older RFC, at least with regard to acceptance testing?

Yes, basically #119 is too large and too sprawling to _actually_ implement.  I started chopping it up into smaller more bite sized RFCs (like #232 and this one) so that we can make incremental progress forward.  The main thing left in #119 that still is not addressed by either #232 or this RFC is the auto-discovery of test helpers and hooks ([this section](https://github.com/rwjblue/rfcs/blob/42/text/0000-grand-testing-unification.md#general-hooks)).

rwjblue commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+  * setup `this.visit` method to visit the given url
+  * setup getter for `this.currentRouteName` which returns the current route name
+  * setup getter for `this.currentURL` which returns the current URL
+* add DOM interaction helpers (heavily influenced by @cibernox's lovely addon [ember-native-dom-helpers](https://github.com/cibernox/ember-native-dom-helpers))
```
Yes, exactly. My plan (and discussed with @cibernox and @turbo87 a couple of times) is to essentially copy the native DOM implementations from `ember-native-dom-helpers` (the [addons own README](https://github.com/cibernox/ember-native-dom-helpers#ember-native-dom-helpers) infers that it is specifically aiming for this).

rwjblue commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+  * setup getter for `this.currentURL` which returns the current URL
+* add DOM interaction helpers (heavily influenced by @cibernox's lovely addon [ember-native-dom-helpers](https://github.com/cibernox/ember-native-dom-helpers))
+  * setup a getter for `this.element` which returns the DOM element representing
+    the root element
```
@simonihmig - Yes. I will update the RFC though.

Turbo87 commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+* invoke `ember-test-helper`s `setupContext` with the tests context (which does the following):
+  * create an owner object and set it on the test context (e.g. `this.owner`)
+  * setup `this.set`, `this.setProperties`, `this.get`, and `this.getProperties` to
+    the test context
```
why do we need those methods?

rwjblue replied:
@Turbo87:

The main reason `set` / `setProperties` methods exist today, is that sets during testing trigger an assertion if not manually wrapped in a runloop. This is something that we are actively working to remove, but that effort will likely take time (as it essentially requires Ember 3.0 where we can rely on all supported platforms having a microtask queue).  

Once that deficiency is addressed, and we no longer need to use manual runloop wrapping during set operations in testing mode, we should consider deprecating (in favor of simply setting them as appropriate) them...

Turbo87 replied:
thanks for the explanation, but I still don't see in what situations one would use e.g. `this.set()` in an **acceptance** test 🤔

rwjblue replied:
Ahh, I see what you are saying now, sorry about that.  I agree, I don't think we strictly need them (and I suspect 99.9% of tests will not use them anyways). I included them here so that the differences between the various types of tests are as small as possible...

Turbo87 replied:
I think I'd prefer to not include them, so that people don't get the wrong idea about the things that they could use

rwjblue replied:
Seems ok to me. We don't expect folks to need them, seems fine to avoid adding them. Will require some additional changes to the existing system in `ember-test-helpers`, but that's quite simple anyways.

rwjblue commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+  * setup a getter for `this.element` which returns the DOM element representing
+    the root element
+  * if `jQuery` is present in the application sets up `this.$` method to run
+    jQuery selectors rooted to `this.element`
```
I had written a very long explanation about how you were wrong, at the end of which I realized you were right...

Suggesting that folks use `$(selector, this.element)` as needed (and from the codemod) seems absolutely fine to avoid creating a cliff.

NLincoln commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,223 @@
+
+# Alternatives
+
+* Do nothing?
```
An alternative (or maybe just a suggestion?) would be to make `ember-native-dom-helpers` into a semi-official addon. Instead of copying the source code into `ember-test-helpers`, why not start shifting the helper code out of the core and into an addon?

rwjblue replied:
Great question, I will add this to the alternatives section.

`ember-test-helpers` is an addon 😸 (completely separate from Ember's own source), supports many Ember versions simultaneously, is focused on providing consistent testing experiences for Ember users, is _already_ a dependency of every application that uses ember-qunit or ember-mocha (roughly all projects), and is a better name 😈.

cibernox replied:
It is indeed an alternative, although I think it makes more sense to make `ember-test-helpers` to be THE PLACE where all "official" testing stuff (that is not qunit-specific) lives.

cibernox commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,304 @@
+(and implemented in
+[ember-qunit](https://github.com/emberjs/ember-qunit)@3.0.0) will be modified to add the same DOM interaction helpers mentioned above:
+
+* setup `this.click` helper method
```
I still find weird that we have two supported ways of doing the same thing. Also, specially in integration tests, it would be extraordinary that someone could do 
```
this.click = function(e) {
  assert.ok(e, 'The event is fired')
}
await render(hbs`{{my-component action=click}}`)
```

Do you think that supposed improvement in debuggeablility justifies this duality. Couldn't just the docs serve as ... documentation of what helpers exists and how to use them?

Turbo87 replied:
FYI There is one potential advantage of using `await this.click()` over `await click()` in that you could potentially run tests concurrently. The `click()` variant has no reliable way of figuring out which test just called that helper, while `this.click()` is bound to the right context.

rwjblue replied:
> There is one potential advantage of using await this.click() over await click() in that you could potentially run tests concurrently.

Confirm, however I tried to avoid talking about this in #119 (and here) because there are many other issues with running tests in parallel (there is only one DOM, focus and other events don't fire the same, etc) and I didn't want to open a can of worms.

cibernox replied:
Even if I slightly favour the imported helpers, my point is about having both. I think it should be one or the other.

rwjblue replied:
> Do you think that supposed improvement in debuggeablility justifies this duality? Couldn't just the docs serve as ...well... documentation of what helpers exists and how to use them?

Sure, I agree that this is definitely a good suggestion also. I'll add a fleshed out version of it to alternatives.

rwjblue replied:
> Even if I slightly favour the imported helpers, my point is about having both. I think it should be one or the other.
  
Right, totally understood. I guess I'm saying that the fact that there are both is an implementation detail. The primary public API and what blueprints will use will be the importable versions. The fact that they desugar to invoking methods on the test context is a minor detail (and _might_ enable features like what @Turbo87 alludes to in https://github.com/emberjs/rfcs/pull/268#discussion_r149123309).

simonihmig replied:
Agree. Made the same point [here](https://github.com/emberjs/rfcs/pull/268#discussion_r149086095). @rwjblue just added some clarification [here](https://github.com/emberjs/rfcs/pull/268/commits/4bb68572760b76d94b19809d0da84d186e05ec8b), which seems reasonable. I would not see it as a problem to have both, as long as there is *one* ideomatic way to write your tests, that is consistently used (guides, blueprints, codemods etc.)

rwjblue replied:
Oh, y'all did it now.  Just removed the `this.foo` versions from the RFC.  I agree having two different public API's for the same thing is confusing, and reserving a bunch of names on the test context is annoying and error prone for users. 

It is still likely that during implementation I will have _some_ mechanism to discover and access these helpers on the test context but it does not need to be public API at all...

rtablada commented:
@rwjblue I like the idea of moving the functions from `ember-qunit` to `ember-test-helpers`.

I would mention that the summary does a good job of describing motivations, but doesn't really give a top level explanation of what changes are being proposed and assumes that contributors are deeply in the know on testing unification and prior RFCs.
I think we should add a sentence of the actual impact on how this moves the needle forward.

Turbo87 commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,272 @@
+// **** after ****
+import { module, test } from 'qunit';
+import { setupAcceptanceTest } from 'ember-qunit';
+import { visit, fillIn, click } from 'ember-test-helpers';
```
as mentioned in https://github.com/emberjs/ember-test-helpers/issues/240 we might want to move to `@ember/test-helpers` for the import paths

rwjblue replied:
Good point, updated in ebfdbe1.

rwjblue commented:
@rtablada 

> assumes that contributors are deeply in the know on testing unification and prior RFCs

I was trying to avoid this by cross-link to any prior work that I referred to, did I miss some references?

> I think we should add a sentence of the actual impact on how this moves the needle forward.

Sounds good, I've tried to do a better job and pushed those changes in b05ed2c. Mind reviewing?

rtablada commented:
@rwjblue seems like the main big changes are:

1. Moving helpers from `ember-qunit` to `ember-test-helpers`
2. Favor `async`/`await`
3. Change to typescript in the dead of night :P

rwjblue commented:
> Change to typescript in the dead of night :P

No one is suggesting this.

rtablada commented:
I was kidding. But the `declare module` syntax may be a bit confusing to people.

rwjblue commented:
@rtablada 

> Moving helpers from ember-qunit to ember-test-helpers

This isn't actually true. Nothing that `ember-qunit` is currently providing is actually moving into `ember-test-helpers`, what am I missing?

> Favor async/await

Indeed, but this isn't really the point at all. If an app doesn't want to use async/await they can happily continue to use promises. This is definitely not "summary" material IMHO.

rwjblue commented:
> the declare module syntax may be a bit confusing to people

Yeah, I was worried about that.  I was trying to find a way to clearly define the **exact** API (inputs, outputs, etc), so I just lifted TypeScript syntax and added some inline docs for prose.  Happy to tweak to another format, but I am pretty opposed to "just describe the API in prose" because I personally find that extremely hard to read and understand.

rtablada commented:
I think just removing the wrapping `declare module 'ember-test-helpers'` still tells the full story.
Maybe switching to comments for typehints?

rtablada commented:
> Nothing that ember-qunit is currently providing is actually moving into ember-test-helpers, what am I missing?

I think it's still unclear what the actual being proposed is then.

[In this section](https://github.com/emberjs/rfcs/blob/acceptance-testing/text/0268-acceptance-testing-refactor.md#dom-interaction-helpers) it seems like you are proposing to move the helpers into the `ember-test-helpers` package/module.

rtablada commented:
IMO moving: https://github.com/emberjs/rfcs/blob/acceptance-testing/text/0268-acceptance-testing-refactor.md#changes-from-current-system higher up will make it easier for people to understand the impact before looking at the pseudo implementation.

rwjblue commented:
> In this section it seems like you are proposing to move the helpers into the ember-test-helpers package/module.

Yep, exactly! But they aren't moving from `ember-qunit` (they were never there). They were previously provided by an asset generated by Ember itself.

rwjblue commented:
> IMO moving: https://github.com/emberjs/rfcs/blob/acceptance-testing/text/0268-acceptance-testing-refactor.md#changes-from-current-system higher up will make it easier for people to understand the impact before looking at the pseudo implementation.

I actually find that having a general before and after snippet (even if not all of the details are obvious) provides a much better frame of reference for reading the rest of the prose. Not sure how to square the two opinions here... 🤔

cibernox commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,278 @@
+  * setup importable `focus` helper method
+  * setup importable `blur` helper method
+  * setup importable `waitUntil` helper method
+  * setup importable `waitFor` helper method
```
- `waitFor` doesn't exist in `ember-native-dom-helpers`. So you like to have it added prior to the migration?

cibernox replied:
I see that you haven't included `find` and `findAll` intentionally, with the idea that that users can just "use the platform". However, specially `find` compared to `this.element.querySelector()` is much more terse. Same goes for `findAll` vs `this.element.querySelectorAll()`, with the extra annoyance that `querySelectorAll` returns a `NodeList` instance instead of an Array, which is cumbersome to use to say the list.

I'd consider adding `find` and `findAll`. I believe that people will end up using them anyway.

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,278 @@
+  * `tests/helpers/resolver.js`
+  * `tests/helpers/start-app.js`
+  * `tests/helpers/destroy-app.js`
+  * `tests/helpers/module-for-acceptance.js`
```
Does tests/test-helper.js remain?

rwjblue replied:
@kategengler - Yes, though IMHO, it would be better to rename it to `tests/index.js`.  It truly is the entry point to all the tests (and even more obvious now that you have to manually call `start()` to kick off the tests)...

0xadada replied:
Many folks use `module-for-acceptance` to start/stop Mirage servers. So the new blueprint wont include that module helper, but nothing is stopping users from keeping or adding it back, correct?

Turbo87 replied:
with the new system you would call something like `setupMirage(hooks)` and it would handle starting and stopping of the server in the right hooks automatically for you

kategengler commented:
How will an addon be able to provide helpers to both the old and new testing systems? Maybe, for an example, the `withFeature` helper from ember-feature-flags:

```
import Ember from 'ember';

export function withFeature(app, featureName) {
  let featuresService = app.__container__.lookup('service:features');
  featuresService.enable(featureName);
}

Ember.Test.registerHelper('withFeature', withFeature);
```

cibernox commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,278 @@
+/**
+  Returns a promise which resolves when the provided callback returns a truthy value.
+*/
+export function waitUntil(() => boolean): Promise<void>;
```
`export function waitUntil(() => boolean, { timeout = 1000 }): Promise<void>;`

cibernox commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,278 @@
+/**
+  Returns a promise which resolves when the provided selector (and count) becomes present.
+*/
+export function waitFor(selector: string, count?: number): Promise<void>;
```
`export function waitFor(selector: string, { count?: number }): Promise<void>;`

rwjblue replied:
Great catch, updated!

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,278 @@
+  that FastBoot itself uses to avoid running initializers for each acceptance
+  test.
+* Implicit promise chaining will no longer be present. If your test needs to
+  wait for a given promise, it should use `await`.
```
I think it is worth noting that `await`ed helpers in acceptance tests will wait for all the same things they currently do.

rwjblue commented:
@kategengler RE: https://github.com/emberjs/rfcs/pull/268#issuecomment-343542474

Assuming the following input:

```js
import Ember from 'ember';

export function withFeature(app, featureName) {
  let featuresService = app.__container__.lookup('service:features');
  featuresService.enable(featureName);
}

Ember.Test.registerHelper('withFeature', withFeature);
```

In order for an addon to support both the existing acceptance testing system, and the new system it could replace that helper with the following:

```js
import { registerAsyncHelper } from '@ember/test';

export function enableFeature(owner, featureName) {
  let featuresService = this.owner.lookup('service:features');
  featuresService.enable(featureName);
}

export default function setupFeatures(hooks, featureName) {
  hooks.beforeEach(function() {
    setupFeature(this.owner, featureName);
  })
}

registerAsyncHelper('withFeature', function(app, featureName) {
  setupFeature(app.__container__, featureName);
});
```

This allows both the prior API (without modification) and the following:

```js
// Option 1:
import { module, test } from 'qunit';
import { setupAcceptanceTest } from 'ember-qunit';
import { setupFeature } from 'addon-name-here/test-support';

module('asdf', function(hooks) {
  setupAcceptanceTest(hooks);
  setupFeature(hooks, 'feature-name-here');

  test(...);
});

// Option 2:
import { module, test } from 'qunit';
import { setupAcceptanceTest } from 'ember-qunit';
import { enableFeature } from 'addon-name-here/test-support';

module('asdf', function(hooks) {
  setupAcceptanceTest(hooks);

  test('awesome test title here', function(assert) {
    enableFeature(this.owner, 'feature-name-here');

    // ...snip...
  });
});
```

(I added ^ to the RFC)

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,370 @@
+}
+
+registerAsyncHelper('withFeature', function(app, featureName) {
+  setupFeature(app.__container__, featureName);
```
I think this is meant to call `enableFeature`

rwjblue replied:
DOH!  Fixed (both)!

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,370 @@
+import { registerAsyncHelper } from '@ember/test';
+
+export function enableFeature(owner, featureName) {
+  let featuresService = this.owner.lookup('service:features');
```
I think this line is meant to use the passed in `owner`

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,370 @@
+
+export default function setupFeatures(hooks, featureName) {
+  hooks.beforeEach(function() {
+    setupFeature(this.owner, featureName);
```
I think this is meant to call `enableFeature`

kategengler commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,370 @@
+
+module('asdf', function(hooks) {
+  setupAcceptanceTest(hooks);
+  setupFeature(hooks, 'feature-name-here');
```
setupFeatures?

rwjblue replied:
I renamed the example test helper to be singular (because it currently only accepts a single feature name).

rwjblue commented:
This was discussed at todays Ember core team meeting, and we believe this is ready to move into the final comment period.

For those that haven't reviewed it yet, now's a great time!

ro0gr commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,352 @@
+# Alternatives
+
+* Do nothing?
+* Make `ember-native-dom-helpers` a default addon (removing the need for DOM interaction helpers proposed here).
```
Looks like `ember-test-helpers` with the `setupTest`, `moduleFor*` and company is already quite complicated.
Keeping DOM helpers implementation as a separate package seems like a win to me from the separation of concerns standpoint.

rwjblue replied:
@ro0gr - Forcing users to be aware of the internal packaging concerns is not ideal. From their perspective, they are using a single set of cohesive tools for testing an Ember app. Whether the implementation actually lives in `ember-test-helpers` or a new addon that is depended on by (and re-exported from) `ember-test-helpers` is not exactly important. 

Also, I disagree that `ember-test-helpers` is "too big".The non-legacy portions are actually quite small at 423 lines of code (even including the legacy code its ~ only 1353 loc).

jamesarosen commented:
We’ve been using some of these patterns in our projects and we really like them.

The global asynchronous test helpers caused problems because we would inadvertently mix synchronous and asynchronous code, breaking tests. We changed from those to chained helpers, which were safer, but unwieldy. Using `async/await` has been a big win.

We’ve also been switching more of our tests to use ember-native-dom-helpers. These are easier for people who have JS/DOM background, but less Ember experience.

Lastly, we’ve been migrating many of our acceptance tests to integration tests, which the above two techniques make easier. Acceptance tests are slower and more brittle. They’re quite valuable, so we do still use them; I’m just saying that these new patterns allow us to write integration tests with much of the power previously reserved for acceptance tests.

Big +1 to the general approach outlined

cibernox commented on text/0268-acceptance-testing-refactor.md:
```diff
@@ -0,0 +1,352 @@
+/**
+  Returns a promise which resolves when the provided callback returns a truthy value.
+*/
+export function waitUntil(() => boolean, { timeout = 1000 }): Promise<void>;
```
I realied now that, in its current implementation, `waitUntil` (and `waitFor`) do not return `Promise<void>`. The promise that returns from `waitUntil` resolves with the resolve value of the callback. The `waitFor` helper does the same, meaning that it will return a promise that resolves to one (or many) DOM elements.

I did this as a convenience for developers wanting to do assertions on the element they've been waiting for.
Example:

```js
let loadingEl = await waitFor('.loading-msg');
assert.equal(loadingEl.textContent.trim(), 'Please, wait...');
```

I am not opposed to change it, but if you think it can be handy, please, update the signature.

rwjblue replied:
👍 updated

rwjblue commented:
Made a few tweaks since my last comment:

* Updated typings for `waitFor` and `waitUntil` (based on @cibernox's feedback)
* Renamed `setupAcceptanceTest` to `setupApplicationTest`. The primary motivation here is that the method name now describes _what_ it does **not** why it is used. For example, `setupRenderingTest` is what we invoke when we will be rendering a template snippet.

raido commented:
This is awesome. 🎉

rwjblue commented:
This was discussed at the Ember core team meeting today, and given that no objections / suggestions / tweaks have come up during the final comment period, and it seems that most folks massively in favor of this improvement we are going to land....

FWIW, I have begun implementation and hope to be able to use this in apps soon...
